### PR TITLE
User story 30

### LINKED ISSUE #31: User Story 30
Image

As an admin,
When I visit the admin merchants index (/admin/merchants)
Then I see the names of the top 5 merchants by total revenue generated
And I see that each merchant name links to the admin merchant show page for that merchant
And I see the total revenue generated next to each merchant name

Notes on Revenue Calculation:
- Only invoices with at least one successful transaction should count towards revenue
- Revenue for an invoice should be calculated as the sum of the revenue of all invoice items
- Revenue for an invoice item should be calculated as the invoice item unit price multiplied by the quantity (do not use the item unit price)

### PR DESCRIPTION
**Pull request by awiley33**
# Describe the changes below:
Added model class method to Merchant. Added display logic to admin/merchants/ view. Feature and model testing.
# Relevant user story:
As an admin,
When I visit the admin merchants index (/admin/merchants)
Then I see the names of the top 5 merchants by total revenue generated
And I see that each merchant name links to the admin merchant show page for that merchant
And I see the total revenue generated next to each merchant name

Notes on Revenue Calculation:

Only invoices with at least one successful transaction should count towards revenue
Revenue for an invoice should be calculated as the sum of the revenue of all invoice items
Revenue for an invoice item should be calculated as the invoice item unit price multiplied by the quantity (do not use the item unit price)

# Before submitting, check the following:
- [ ] Entire test suite is passing **1 failing test, Weston has since fixed this**
- [99.82] SimpleCov test percentage

# Note any co-authors / anything else you'd like to add:
I have been working on this branch for a while, therefore it is very out of date with main. I can handle any merge conflicts

### CODE REVIEW

File: spec/features/admin/merchants/index_spec.rb [hunk 1]
```diff
@@ -8,11 +8,68 @@
     @merchant_4 = Merchant.create!(name: "Cummings-Thiel")
     @merchant_5 = Merchant.create!(name: "Williamson Group")
     @merchant_6 = Merchant.create!(name: "Bernhard-Johns")
-  end
+    @merchant_7 = Merchant.create!(name: "Anna")
+    @merchant_8 = Merchant.create!(name: "Luca")
```
Comment (dani-wilson) [inline spec/features/admin/merchants/index_spec.rb:12]: the bestest merchant that ever lived 🐕

**Review by dani-wilson — APPROVED** [review]
Awesome work!!!

File: app/controllers/admin/merchants_controller.rb [hunk 2]
```diff
@@ -26,8 +26,25 @@ def edit
...
+
+  def create
+    @merchant = Merchant.new(new_merchant_params)
+    if @merchant.save
+      redirect_to admin_merchants_path
+    else
+      flash[:notice] = "Please enter a merchant name to continue."
+      render "new"
+    end
+  end
```
Comment (westonio) [inline app/controllers/admin/merchants_controller.rb]: Thanks for covering this sad path!

File: app/models/merchant.rb [hunk 3]
```diff
@@ -13,4 +13,13 @@ def favorite_customers
     ids = invoices.pluck(:id)   
     Customer.select("first_name, last_name, count(result)").joins(invoices: :transactions).where(transactions: {result: "success"}).where(invoices: {id: ids}).order(count: :desc).group("first_name, last_name").limit(5)
   end
-end
+
+  def self.top_merchants
+    Merchant.select('merchants.name, merchants.id, SUM(invoice_items.unit_price * invoice_items.quantity) AS total_revenue')
+    .joins(items: { invoice_items: { invoice: :transactions } })
```
Comment (westonio) [inline app/models/merchant.rb:20]: So cool seeing a joins like this!

**Review by westonio — APPROVED** [review]

**Pull request merged**